### PR TITLE
Add requirements for JWT support

### PR DESCRIPTION
**Pull request by paddybyers**
Note we still need to finalise the link to the documentation that states what requirements a JWT must satisfy.

In all cases the library must handle token strings opaquely and no library functionality handles JWT token strings differently from Ably token strings.

The additions in this PR are to:
- define the method to determine whether a string is an API key or a token string;
- define token string;
- clarify which cases need to be tested;
- add explicit support for the `application/jwt` MIME type in an `authURL` response.

### CODE REVIEW

File: content/client-lib-development-guide/features.textile [hunk 1]
```diff
@@ -56,7 +56,10 @@ h2(#rest). REST client library
 
 h3(#restclient). RestClient
 
-* @(RSC1)@ The constructor accepts either an API key, a token string, or a set of "@ClientOptions@":#options.  If invalid arguments are provided such as a no API key, no token and no means to create a token, then this will result in an error
+* @(RSC1)@ The constructor accepts a set of "@ClientOptions@":#options or, in languages that support overloaded constructors, a string which may be a token string or an API key.
+** @(RSC1a)@ If a single string argument is supplied when constructing the library then the library must determine whether this is a key or a token by checking for the presence of the ':' (colon) delimiter present in an API key. Any other string must be treated as a token string.
+** @(RSC1b)@ If invalid arguments are provided such as a no API key, no token and no means to create a token, then this will result in an error
```
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: typo: "provided such as [a] no API key"

Comment (mattheworiordan) [reply content/client-lib-development-guide/features.textile]: Should we be more explicit about the error message, code etc?

**Comment by funkyboy** [comment]
@mattheworiordan Yes:
- https://github.com/ably/ably-java/issues/384
- https://github.com/ably/ably-js/issues/505
- https://github.com/ably/ably-ruby/issues/136
- https://github.com/ably/ably-ios/issues/713
- https://github.com/ably/ably-dotnet/issues/210
- https://github.com/ably/ably-python/issues/108

**Comment by mattheworiordan** [comment]
@funkyboy please see https://github.com/ably/echoserver for the `echoserver` I mentioned.  You should add JWT functionality in there to provide permutations of tokens such as an Ably JWT token, an embedded Ably token in a JWT etc.  I would recommend that the token server has no knowledge of API keys and instead you just pass an API key / token to it in the URL, and it in turn uses that API key to generate a JWT.  I have also added you to the app hosted in Heroku.

**Comment by funkyboy** [comment]
Just noticed a TODO here https://github.com/ably/docs/pull/420/files#diff-11900b395df266bc2bbfc1d11bdfc7a0R1140

**Comment by funkyboy** [comment]
@mattheworiordan @SimonWoolf @paddybyers can we merge this? 
We need labels for the JWT spec items so we can add them to the spreadsheet and use them to tag tests.

**Comment by paddybyers** [comment]
> Just noticed a TODO here

@tomczoink can you provide the URL please?

**Comment by tomczoink** [comment]
@paddybyers link for Ably requirements for JWT is https://www.ably.io/documentation/general/authentication#ably-jwt

**Comment by paddybyers** [comment]
I'm happy for this to be merged. @SimonWoolf @mattheworiordan ?

File: content/client-lib-development-guide/features.textile [hunk 2]
```diff
@@ -56,7 +56,10 @@ h2(#rest). REST client library
 
 h3(#restclient). RestClient
 
-* @(RSC1)@ The constructor accepts either an API key, a token string, or a set of "@ClientOptions@":#options.  If invalid arguments are provided such as a no API key, no token and no means to create a token, then this will result in an error
+* @(RSC1)@ The constructor accepts a set of "@ClientOptions@":#options or, in languages that support overloaded constructors, a string which may be a token string or an API key.
+** @(RSC1a)@ If a single string argument is supplied when constructing the library then the library must determine whether this is a key or a token by checking for the presence of the ':' (colon) delimiter present in an API key. Any other string must be treated as a token string.
+** @(RSC1b)@ If invalid arguments are provided such as no API key, no token and no means to create a token, then this will result in an error with error code 40000 and an informative message.
```
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: @paddybyers should we not consider a new error code here?  I believe the goal with 1.1 is to provide useful error codes with URLs to specific advice on how to fix them, whereas 40000 is just a generic code.

**Comment by mattheworiordan** [comment]
@paddybyers @funkyboy happy to be merged too, although I'd prefer a custom error code instead of 40000, in the interests of developer experience once 1.1 is released

File: content/client-lib-development-guide/features.textile (hunk 2 above)
Comment (mattheworiordan) [reply content/client-lib-development-guide/features.textile]: > Unable to obtain credentials from given parameters

Sure, that makes sense to me

File: content/client-lib-development-guide/features.textile [hunk 3]
```diff
@@ -122,7 +126,7 @@ h3(#rest-auth). Auth
...
-** @(RSA4f)@ The following conditions imply that the token is in an invalid format: the @authUrl@ response content type is neither @text/plain@ nor @application/json@; the object passed by @authCallback@ is neither a @String@, @JsonObject@, @TokenRequest@ object, nor @TokenDetails@ object; the text token string is greater than 384 bytes; the JSON stringified @JsonObject@, @TokenRequest@ or @TokenDetails@ is greater than 128kb.
+** @(RSA4f)@ The following conditions imply that the token is in an invalid format: the @authUrl@ response content type is neither @text/plain@ nor @application/json@; the object passed by @authCallback@ is neither a @String@, @JsonObject@, @TokenRequest@ object, nor @TokenDetails@ object; the token string or the JSON stringified @JsonObject@, @TokenRequest@ or @TokenDetails@ is greater than 128kb.
```
Comment (funkyboy) [inline content/client-lib-development-guide/features.textile]: > neither text/plain nor application/json

@paddybyers It's mentioned in RSA8c below but I think we should mention `application/jwt` also here

File: content/client-lib-development-guide/features.textile (hunk 2 above)
Comment (paddybyers) [reply content/client-lib-development-guide/features.textile]: https://github.com/ably/ably-common/pull/28

**Comment by paddybyers** [comment]
I've added https://github.com/ably/docs/pull/420/commits/a5abc73ff3453348cd3f2bcd4855bad011a236a5.

I'll merge unless anyone objects @mattheworiordan @funkyboy @SimonWoolf @tomczoink

**Comment by mattheworiordan** [comment]
Sure, merge away 👍

**Pull request merged**